### PR TITLE
ICU-22453 Fix non null terminated buffer issue.

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22453
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [X] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/test/intltest/uniquecharstrtst.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    int32_t index2 = strings.add(key2, status);
+    if (U_FAILURE(status)) {
+       errln("Cannot add key2 - %s", u_errorName(status));
+    }
+    strings.add(key3, status);
+    if (U_FAILURE(status)) {
+       errln("Cannot add key3 - %s", u_errorName(status));
+    }
+    strings.freeze();
+    if (strcmp(strings.get(index1), "abcde") != 0) {
```
Comment (richgillam) [inline icu4c/source/test/intltest/uniquecharstrtst.cpp:49]: You could just use `assertUEquals()` here (and below), couldn't you?

**Review by richgillam — APPROVED** [review]
LGTM.

File: icu4c/source/common/uniquecharstr.h [hunk 2]
```diff
@@ -62,7 +62,10 @@ class UniqueCharStrings {
             return 0;
         }
         // The string points into the resource bundle.
-        const char16_t *p = s.getBuffer();
+        // s.getBuffer() may return not null-terminated string
+        // and casue illegal memory access problem because
+        // uhash_geti/puti expect null-terminated string.
+        const char16_t *p = ((UnicodeString&)s).getTerminatedBuffer();
```
Comment (sffc) [inline icu4c/source/common/uniquecharstr.h:68]: Issue: This cast changes the constness of `s`. You should either change the function to take a non-const `s` or clone `s` before getting the terminated buffer.

**Review by sffc — CHANGES_REQUESTED** [review]

**Comment by FrankYFTang** [comment]
somehow the PR does not getting my latest force push to https://github.com/FrankYFTang/icu/tree/ICU-22453-string-add See https://github.com/FrankYFTang/icu/commit/12e69db28888f6f3f94737f5f1f7c80c7e4d566b

**Comment by FrankYFTang** [comment]
this PR is strange. I gave up. Create another one in https://github.com/unicode-org/icu/pull/2543. and clsoe this one

**Pull request closed without merging**